futures-select-macro/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+        }
+
+        while !input.is_empty() {
+            let case_kind = if input.peek(kw::complete) {
+                // `complete`
+                if select.complete.is_some() {
+                    return Err(input.error("multiple `complete` cases found, only one allowed"));
+                }
+                input.parse::<kw::complete>()?;
+                CaseKind::Complete
+            } else if input.peek(Token![default]) {
+                // `default`
+                if select.default.is_some() {
+                    return Err(input.error("multiple `default` cases found, only one allowed"));
+                }
+                input.parse::<Ident>()?;
+                CaseKind::Default
+            } else {
+                // `<pat> = <expr>`
+                let pat = input.parse()?;
+                input.parse::<Token![=]>()?;
+                let expr = input.parse()?;
+                CaseKind::Normal(pat, expr)
+            };
+
+            // `=> <expr>`
+            input.parse::<Token![=>]>()?;
+            let expr = input.parse::<Expr>()?;
+
+            // Commas after the expression are only optional if it's a `Block`,
+            // otherwise they are required.
+            match &expr {
+                Expr::Block(_) => { input.parse::<Option<Token![,]>>()?; }
+                _ => { input.parse::<Token![,]>()?; }
+            }
+
+            match case_kind {
+                CaseKind::Complete => select.complete = Some(expr),
+                CaseKind::Default => select.default = Some(expr),
+                CaseKind::Normal(pat, fut_expr) => {
+                    select.normal_fut_exprs.push(fut_expr);
+                    select.normal_fut_handlers.push((pat, expr));
+                },
+            }
+        }
+
+        Ok(select)
+    }
+}
+
+// Enum over all the cases in which the `select!` waiting has completed and the result
+// can be processed.
+//
+// `enum __PrivResult<_1, _2, ...> { _1(_1), _2(_2), ..., Complete }`
+fn declare_result_enum(
+    result_ident: Ident,
+    variants: usize,
+    complete: bool,
+    span: Span
+) -> (Vec<Ident>, syn::ItemEnum) {
+    // "_0", "_1", "_2"
+    let variant_names: Vec<Ident> =
+        (0..variants)
+            .map(|num| Ident::new(&format!("_{}", num), span))
+            .collect();
+
+    // `<_1, _2, ...>`
+    let enum_generics = variant_names.iter()
+        .map(|name| {
+            syn::GenericParam::Type(parse_quote!( #name ))
+        })
+        .collect();
+
+    let unit_variant = |name| syn::Variant {
+        attrs: vec![],
+        ident: Ident::new(name, span),
+        fields: syn::Fields::Unit,
+        discriminant: None,
+    };
+
+    // `_1(_1), _2(_2), ..., Complete`
+    let enum_variants = variant_names.iter()
+        .map(|name| parse_quote!( #name ( #name ) ))
+        .chain(if complete { Some(unit_variant("Complete")) } else { None })
+        .collect();
+
+    (
+        variant_names,
+        syn::ItemEnum {
```

Comment:
In my code I typically avoid constructing Syn types. Syn's parsers construct Syn types while the macro only constructs token streams. That might look something like this:

```rust
// "_0", "_1", "_2"
let variant_names: Vec<Ident> =
    (0..variants)
        .map(|num| Ident::new(&format!("_{}", num), span))
        .collect();

let type_parameters = &variant_names;
let variants = &variant_names;

let complete_variant = if complete {
    Some(quote!(Complete))
} else {
    None
};

let enum_item = quote! {
    enum #result_ident<#(#type_parameters,)*> {
        #(
            #variants(#type_parameters),
        )*
        #complete_variant
    }
};

(variant_names, enum_item)
```

Up to you which you prefer, I am also fine with your current way.

Reply:
Oh this is much better, thank you!